app/src/main/java/org/sil/gatherwords/SessionActivity.java

```diff
@@ -118,18 +129,127 @@ public void save_settings_fab_pressed(View view) {
         startActivity(i);
     }
 
-    private static class InsertSessionsTask extends AsyncTask<Session, Void, Void> {
+    private static class InsertSessionsTask extends AsyncTask<Session, Double, Void> {
         private SessionDao sDAO;
+        private WordDao wordDao;
+        // TODO: fix null object reference so we don't risk a data leak with the context
+//        private final ThreadLocal<SessionActivity> sessionActivity = new ThreadLocal<>();
+        private Long sessionID;
+        private int maxProgress;
+        private AppDatabase db;
+        private String wordList;
+        private final Context context;
 
-        InsertSessionsTask(AppDatabase db) {
+        InsertSessionsTask(SessionActivity activity) {
+//            sessionActivity.set(activity);
+//            db = AppDatabase.get(sessionActivity.get().getApplicationContext());
+            context = activity.getApplicationContext();
+            db = AppDatabase.get(context);
+            wordList = activity.worldListToLoad;
+            wordDao = db.wordDao();
             sDAO = db.sessionDao();
         }
 
         @Override
         protected Void doInBackground(Session... sessions) {
-            sDAO.insertSession(sessions);
+            List<Long> ids = sDAO.insertSession(sessions);
+
+            if (ids.get(0) != null) {
```

Comment:
This will crash if no sessions where inserted.

Reply:
I'll fix it, but when would this line run if there are no functions. The only entry point creates the session being inserted